File: problems/42/solution.cpp
// 	 struct TreeNode {
// 	     string val;
// 	     TreeNode* left;
//       TreeNode* right;
//       TreeNode(string x): val(x), left(NULL), right(NULL) {}
//   };
//
//   class Solution {
// 	   public:
// 		  vector<vector<int>> zigzagLevelOrder(TreeNode* root);
//   };
//
//   Please refrain from using print statements

vector<vector<int>> Solution::zigzagLevelOrder(TreeNode* root) {
    vector<vector<int>> result;
    if (!root) {
        return result;
    }
    stack<TreeNode*> S1, S2;
    S1.push(root);
	int level = 0;
    while (!S1.empty()) {
        vector<int> row;
		level++;
        while (!S1.empty()) {
            auto currentNode = S1.top(); S1.pop();
            row.push_back(currentNode->val);
			if (level % 2 == 1) {
	            if (currentNode -> left)
	                S2.push(currentNode->left);
	            if (currentNode -> right)
	                S2.push(currentNode -> right);
			} else {
				if (currentNode -> right)
					S2.push(currentNode -> right);
				if (currentNode -> left)
					S2.push(currentNode->left);
			}
        }
        S1 = S2;
        S2 = stack<TreeNode*> ();
        result.push_back(row);
    }
    return result;

}

File: problems/42/solution.java
// TreeNode type
// class TreeNode {
//     int val;
//     TreeNode left;
//     TreeNode right;
//     TreeNode(int x) { val = x; left = null; right = null; }
// }
class Solution {
    public ArrayList<ArrayList<Integer>> zigzagLevelOrder(TreeNode root) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        ArrayList<TreeNode> Q1 = new ArrayList<TreeNode>();
        if (root == null)
            return result;
        int level = 0;
        Q1.add(root);
        while (Q1.size() > 0) {
            ArrayList<TreeNode> Q2 = new ArrayList<TreeNode>();
            ArrayList<Integer> row = new ArrayList<Integer> ();
            int index = 0;
            level++;

            if (level % 2 == 1) {
                index = Q1.size() - 1;
                while (index >= 0) {
                    TreeNode current = Q1.get(index);
                    index--;
                    row.add(current.val);
                    if (current.left != null)
                        Q2.add(current.left);
                    if (current.right != null)
                        Q2.add(current.right);
                }
            } else {
                index = Q1.size() - 1;
                while (index >= 0) {
                    TreeNode current = Q1.get(index);
                    index--;
                    row.add(current.val);
                    if (current.right != null)
                        Q2.add(current.right);
                    if (current.left != null)
                        Q2.add(current.left);
                }

            }
            Q1 = Q2;
            result.add(row);
        }
        return result;
    }

}
